tests/e2e_tests

```diff
@@ -36,14 +37,33 @@ do
         echo "" >> ../error.log
     fi
     cd ..
-done
 
-if [ $failure = "true" ]
-then
-    echo "One of the e2e tests failed, check the logs";
-    cat error.log
-    exit 1
-fi
+    if [ "$failure" = "true" ]
+    then
+        exit 255
+    fi
+}
+
+exitOnError() {
+  exit_code=$1
+  msg=$2
+  if [ "$exit_code" -ne 0 ]; then
+    echo -e "\033[31mEXIT $exit_code $msg\033[0m" >&2
+    # If any invocation of the command exits with a status of 255, xargs will stop immediately without reading any further input.
+    # An error message is issued on stderr when this happens.
+    exit 255
+  fi
+}
+
+dirs=$(find . -maxdepth 1 -mindepth 1 -type d)
+
+# Export the function so xargs can use it
+export -f run_infection_for_dir
+
+# Use xargs to run the run_infection_for_dir function in parallel
+echo "$dirs" | xargs -I {} -P 4 bash -c "run_infection_for_dir {} ${1:-bin/infection}"
```

Comment:
TIL, I thought you had to use GNU parallel, turns out xargs does it just fine, cool!

I do wonder if we could use the makefile instead at some point (also makes it easier to scale up the number of processes), but that's an entirely different set of issue